### ISSUE TITLE
Lifecycle issues on dynamic widget change

### ISSUE THREAD

**Issue by raphlinus** [linebender/druid#1258]
This is a crash when using master Druid with the Crochet prototype. The lifecycle problems were happening before, but only giving warnings. Now text layout is highly dependent on getting lifecycle right. Though it's observed in Crochet, I'm pretty sure it's a problem in Druid, and will probably bite us any time we're trying to reconfigure children dynamically.

The following println-debugging trace should give a clue what's going wrong. It's from a simplified version of the "counter" example, stripped down to just be a label. I can post a gist of the example, and a diff adding the println, if it'd be useful.

```
anywidget lifecycle WidgetAdded
anywidget event WindowConnected
children changed
anywidget layout
anywidget layout
label layout
anywidget lifecycle Size(4.0W├ù16.0H)
label lifecycle Size(4.0W├ù16.0H)
anywidget lifecycle Size(1427.0W├ù726.0H)
anywidget lifecycle Internal(RouteWidgetAdded)
anywidget lifecycle WidgetAdded
label lifecycle WidgetAdded
anywidget paint
anywidget paint
label paint
thread 'main' panicked at 'TextLayout::draw called without rebuilding layout object. Text was 'Foo'', C:\Users\raph\github\druid\druid\src\text\layout.rs:291:9
```

The "children changed" log is when the child widget (a label) gets added to the (modified) flex container; it's [this line](https://github.com/raphlinus/crochet/blob/5043f3283429e5ff0c511b2718523a25f9156ad2/src/widget/flex.rs#L588) in particular.

The underlying problem seems pretty simple: it's doing the WidgetAdded lifecycle *after* the layout, which seems pretty clearly wrong to me. I think this problem may be hidden in other examples because there might be an intervening `update` or something, but I also wouldn't be surprised if it were possible to trigger in extant code even outside Crochet.

I haven't investigated the lifecycle ordering logic carefully yet. It feels like a bit of a haunted graveyard. But I am willing to dig in.

(PS: yes, the × symbol is mojibake. Because it's 2020 and all software is broken)

**Comment by Finnerale** [comment]
This sounds like #1127?
If that is the problem, I've investigated this some time ago and I think I have a patch for this in one of my branches. I'll refresh my memory about this tomorrow and see what I got.

**Comment by raphlinus** [comment]
Yes, on quick skim it seems likely that's the same problem. It would be sweet if there's a patch for it.

**Comment by Finnerale** [comment]
I found the branch and updated it, I'll have to fix some of out tests though because they do not even honor `WidgetAdded`...
For that I'll have to get into that test harness stuff so I probably won't get to finish it today.

**Comment by Finnerale** [comment]
So after 4 to 5 hours of madness I believe I found the reason for this.

First of, if I am correct, this is purely a Crochet issue and not really Druid's fault.

What I believe causes this, is that Crochet's `mutate` method does not use the correct `EventCtx` for the widgets, but instead propagates down the one from the root widget. If you take a look at the `WidgetPod::event` method:
https://github.com/linebender/druid/blob/ecd0a2484fb7d2a809251e08f857baa010ad6723/druid/src/core.rs#L704-L719
you can see, that is uses an `inner_ctx` for the child. Because we don't do this `children_changed` never becomes true for any widgets other than the root, which means that this part of `WidgetPod::lifecycle` always gets skipped:
https://github.com/linebender/druid/blob/ecd0a2484fb7d2a809251e08f857baa010ad6723/druid/src/core.rs#L736-L751
And thus `WidgetAdded` never gets propagated anywhere but to the root.

So far I have no idea how we could do this correctly, because we have no way of constructing such a "child context" from Crochet.

**Comment by raphlinus** [comment]
Thanks, this is a very useful analysis, and not what I was expecting. I haven't thought about it too deeply yet, but it seems like the right path forward might be to add a method to Druid to create an appropriate child context.

**Comment by Finnerale** [comment]
Closed by #1352 and raphlinus/crochet#19 :tada:
